### ISSUE TITLE
multiple threshholds per jail

### ISSUE THREAD

**Issue by jnorell** [fail2ban/fail2ban#623]
I would like to be able to specify a jail configuration with multiple maxretry/findtime/bantime settings to trigger a match.

Currently I find myself repeating sections over and again to cover different rates, as some scans come in very quickly (low maxretry/findtime/bantime), others are low rate and go on for hours/days (higher settings).  Eg. as an example, I have a sasl-30s, sasl-30m and sasl-1d that only differ in those three settings.  They are all useful to catch different scans.

**Comment by leeclemens** [comment]
You can have two jails with different maxretry, bantime and findtime settings already, is that not working?

**Comment by jnorell** [comment]
multiple jails do work, but there currently are performance issues (see #529) and the possibility of one jail removing an ip address which should still be blocked according to another jail's settintgs (see 2nd comment above)

**Comment by leeclemens** [comment]
If you are using pyinotify, I'm not sure how relevant/accurate "It may make sense to move all the time parsing" is. Since you aren't doing any parsing...while two jails would be notified upon changes - it's not like you're reparsing the log file.

**Comment by leeclemens** [comment]
Some iostat info regarding that impact would be helpful, as I haven't seen it be a problem personally.

**Comment by snathan9** [comment]
Just wanted to leave a note that I am in favor of this feature. In my e.g. I have a ban for those hack scripts that are running wild hitting the server many times per second. And then there are scripts that are paused for a couple seconds per hit but still hit the server over a very long period of time. This requires the same config to have different maxretry/findtime/bantime settings. Yes, I am going to repeat the jails but sure would be nice have a good hierarchy. Also makes it cleaner. Probably more efficient as well.

**Comment by sebres** [comment]
> I am going to repeat the jails but sure would be nice have a good hierarchy. Also makes it cleaner.

Do you mean the jail configuration, right?

I'd like to reopen this but not necessarily as an additional threshold probably... Newer version of fail2ban has an observer module, which can react to a finding and a ban, and currently serves only the needs of incremental banning (prolongation) and maxretry reduction for know as bad intruders.
I must think about some new config facilities - how one would be able to specify additional "reactions" causing a ban (e. g. threshold, burst, etc)... Maybe it is much simpler than one can possibly expect.

**Comment by snathan9** [comment]
Yes, I mean't jail configuration. I understand what you mean, but cannot think of how this new config facility may be incorporated. Will leave that up to you :) The more I think about it, it seems easier to just create another jail config section! lol.  In any case, my research let to the idea of creating [sub-sections](https://en.wikipedia.org/wiki/INI_file#Hierarchy_(section_nesting)). This may help you process the set of config files for multiple scenarios within the same block. Hope that helps.

**Comment by snathan9** [comment]
I wanted to add more details and my use case, which I imagine applies to others as well.

I have a Front-facing NGINX reverse proxy that servers traffic to a number of backend applications and websites.  The access logs for each website have been split, for ease of review. I have about 50 log files that NGINX produces - about one for each site.   Hackers are hitting all of these websites and they are all the same culprits so banning them on one bans them on all which is a major win.

Coming to the check itself.  I have the following
(1) Check for any wp-login that results in a 404 and ban immediate (non-wordpress websites)
(2) Check for wp-login that results in a 200, but where this is repeated many times in a short period (wordpress websites)
(3) Check for 404 errors in a short period - e.g. 6 404s in 3 seconds and ban
(4) Check for multiple 404s within in a longer period of time but one a handful per minute. e.g. 30 404's in 3 minutes.

For this I need to have 4 jail configs.  And the jails are scanning 50 log files each time. Ideally I would like to specify a hierarchy as Log file -> Filter -> Reactions. I also have to repeat the .conf file in filter.d two times each with the exact same filter.  

I think at least the definition of the above can be achieved via sub-sections in the configuration. Of course implementation is another story! I also realize some things are easier said but require major architectural change.  Since you are thinking of writing up the new observer module, I wanted to provide you as much feedback. 

Thank you for Fail2ban, much appreciated.

**Comment by sebres** [comment]
related to #529 (#1379)

**Comment by raquelhortab** [comment]
hi! was this ever addressed?
